File: solutions/monotonic-stack/alternatives/316-remove.js
/**

 *
 * This problem demonstrates key concepts in Monotonic Stack.
 *
 * SOLUTION EXPLANATION:
 *
 * INTUITION:
 * We need to remove duplicate letters while maintaining lexicographical order and ensuring each letter
 * appears at least once. A greedy approach with a monotonic stack helps us build the smallest result
 * by removing larger characters when a smaller one can replace them (if duplicates exist later).
 *
 * APPROACH:






 *
 * WHY THIS WORKS:
 * - The stack maintains lexicographically smallest subsequence
 * - We can safely remove a character if it appears later (count > 0)
 * - The "used" set prevents duplicates in result
 * - Greedy removal of larger characters when safe ensures optimal result
 *
 * TIME COMPLEXITY: O(n) - each character processed once
 * SPACE COMPLEXITY: O(k) - where k is unique characters (at most 26 for lowercase letters)
 *
 * EXAMPLE WALKTHROUGH:
 * ```
 * Input: s = "bcabc"
 * Count: {b:2, c:2, a:1}
 *
 * Process 'b': count={b:1,c:2,a:1}, stack=['b'], used={b}
 * Process 'c': count={b:1,c:1,a:1}, stack=['b','c'], used={b,c}
 * Process 'a': 'a'<'c' and count[c]>0, pop 'c'
 *              'a'<'b' and count[b]>0, pop 'b'
 *              stack=['a'], used={a}, count={b:1,c:1,a:0}
 * Process 'b': stack=['a','b'], used={a,b}, count={b:0,c:1,a:0}
 * Process 'c': stack=['a','b','c'], used={a,b,c}
 *
 * Result: "abc"
 * ```
 *
 * EDGE CASES:
 * - Empty string: return ""
 * - No duplicates: return original string
 * - All same character: return single character
 * - Already sorted: return deduplicated string
 */

/**
 * Main solution for Problem 316: Remove Duplicate Letters
 *
 * @param {string} s - Input string
 * @return {string} - Smallest lexicographical string with no duplicates
 *
 * Time Complexity: O(n)
 * Space Complexity: O(k) where k is number of unique characters
 */
function solve(s) {
    if (!s || s.length === 0) return "";

    // Count frequency of each character
    const count = {};
    for (const char of s) {
        count[char] = (count[char] || 0) + 1;
    }

    const stack = [];
    const used = new Set();

    for (const char of s) {
        // Decrement count for current character
        count[char]--;

        // If already in result, skip
        if (used.has(char)) {
            continue;
        }

        // Remove characters that are larger and appear later
        while (stack.length > 0 &&
               stack[stack.length - 1] > char &&
               count[stack[stack.length - 1]] > 0) {
            const removed = stack.pop();
            used.delete(removed);
        }

        // Add current character
        stack.push(char);
        used.add(char);
    }

    return stack.join('');
}

/**
 * Test cases for Problem 316: Remove Duplicate Letters
 */
function testSolution() {
    console.log('Testing 316. Remove Duplicate Letters');

    // Test case 1: Example from problem
    const result1 = solve("bcabc");
    const expected1 = "abc";
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: Another example
    const result2 = solve("cbacdcbc");
    const expected2 = "acdb";
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: Single character repeated
    const result3 = solve("aaaa");
    const expected3 = "a";
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: No duplicates
    const result4 = solve("abc");
    const expected4 = "abc";
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);

    // Test case 5: Reverse order with duplicates
    const result5 = solve("edcba");
    const expected5 = "edcba";
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);

    // Test case 6: Complex case
    const result6 = solve("abacb");
    const expected6 = "abc";
    console.assert(result6 === expected6, `Test 6 failed: expected ${expected6}, got ${result6}`);

    console.log('All test cases passed for 316. Remove Duplicate Letters!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 316. Remove ===');
    console.log('Category: Monotonic Stack');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on monotonic stack concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: docs/solutions/monotonic-stack/316-remove.py
"""
# Difficulty: Medium

# 316. Remove

Given a string s, remove duplicate letters so that every letter appears once and only once. You must make sure your result is the smallest in lexicographical order among all possible results.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>"bcabc"</dd>
<dt>Output:</dt>
<dd>"abc"</dd>
<dt>Explanation:</dt>
<dd>After removing duplicate letters while maintaining lexicographical order, result is 'abc'</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
Use monotonic increasing stack. For each character, while stack top > current character, pop if we can remove it (count > 0 and appears later). This ensures lexicographically smallest result.

### APPROACH:
1. **Count occurrences**: Create counter for each character's frequency
2. **Initialize stack and visited**: stack = [], visited = set()
3. **Iterate through string**: For each char in s
4. **Decrement count**: counter[char] -= 1
5. **Skip if visited**: If char in visited, continue
6. **Maintain order**: While stack and stack[-1] > char and counter[stack[-1]] > 0, pop and remove from visited
7. **Add current**: Append char to stack, add to visited
8. **Return result**: Return ''.join(stack)

### WHY THIS WORKS:
- Monotonic stack builds lexicographically smallest result
- Track remaining count of each char: safe to remove if appears later
- Pop larger chars from stack if count > 0 (can add back later)
- Visited set prevents duplicate characters in result
- Greedy approach works: always try to place smaller char earlier

### EXAMPLE WALKTHROUGH:
```
Input: s = "bcabc"
Step 1: Count frequencies and track remaining
  freq = {'b':2, 'c':2, 'a':1}

Step 2: Build result with monotonic stack
  Add 'b': stack=['b']
  Add 'c': stack=['b','c']
  Add 'a': pop 'c' (a<c, c appears later), pop 'b' (a<b, b appears later)
          stack=['a']
  Add 'b': stack=['a','b']
  Add 'c': stack=['a','b','c']

Output: "abc"
```

### TIME COMPLEXITY:
O(n)

### SPACE COMPLEXITY:
O(1)

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>
"""

class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        """
        Remove duplicate letters to create smallest lexicographical order string.

        Args:
            s: Input string with possible duplicate characters

        Returns:
            Smallest lexicographical string with no duplicate characters

        Time Complexity: O(n) - single pass through string
        Space Complexity: O(1) - stack and sets limited to 26 characters
        """
        # Count occurrences of each character
        last_occurrence = {char: i for i, char in enumerate(s)}

        stack = []
        in_stack = set()

        for i, char in enumerate(s):
            # Skip if character already in result
            if char in in_stack:
                continue

            # Remove larger characters that appear later
            while stack and stack[-1] > char and last_occurrence[stack[-1]] > i:
                removed = stack.pop()
                in_stack.remove(removed)

            stack.append(char)
            in_stack.add(char)

        return ''.join(stack)

    def solve(self, s: str) -> str:
        """Wrapper method for consistency with template."""
        return self.removeDuplicateLetters(s)

def test_solution():
    """
    Test cases for 316. Remove Duplicate Letters.
    """
    solution = Solution()

    # Test case 1: Basic example
    result = solution.solve("bcabc")
    expected = "abc"
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Multiple duplicates
    result = solution.solve("cbacdcbc")
    expected = "acdb"
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 3: No duplicates
    result = solution.solve("abcd")
    expected = "abcd"
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 4: All same character
    result = solution.solve("aaaa")
    expected = "a"
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 5: Reverse alphabetical
    result = solution.solve("dcba")
    expected = "dcba"
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 6: Complex pattern
    result = solution.solve("ecbacba")
    expected = "eacb"
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 7: Single character
    result = solution.solve("a")
    expected = "a"
    assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    s = "bcabc"
    result = solution.solve(s)
    print(f"Solution for 316. Remove Duplicate Letters: {result}")
